### ISSUE TITLE
Windows: px_proxy_factory_get_proxies is very slow

### ISSUE THREAD

**Issue by lovetox** [libproxy/libproxy#99]
Hi,

On my Windows 7 x64 machine this method id very slow, it takes 2,5 seconds to return for https://www.google.com.

```
#include <glib.h>
#include <proxy.h>

int main()
{

  gint64 t1, t2;

  pxProxyFactory *pf = px_proxy_factory_new();
  if (!pf) return 1;

  t1 = g_get_monotonic_time ();
  char **proxies = px_proxy_factory_get_proxies(pf, "https://www.google.com");
  t2 = g_get_monotonic_time ();
  px_proxy_factory_free(pf);

  g_print ("took %f seconds", (gdouble) (t2 - t1) / 1000000);
  return 0;
}
```

Output:
```
$ ./proxy-test.exe
took 2.256647 seconds
```

Any ideas?

**Comment by lovetox** [comment]
I found the debug mode

```
$ _PX_DEBUG=1 ./proxy-test.exe
Configuration extension is: 23w32reg_config_extension
Ignored list is:
Config is:
        wpad://
Trying to find the PAC using WPAD...
WPAD search via: 24dns_alias_wpad_extension
No PAC found...
took 2.253865 seconds
```

It hangs in `WPAD search via: 24dns_alias_wpad_extension`, what is wpad, how do i get rid of it?

**Comment by nirbheek** [comment]
WPAD is this: https://en.wikipedia.org/wiki/Web_Proxy_Auto-Discovery_Protocol

Do you have "automatically detect settings" enabled under Windows Proxy settings?

**Comment by lovetox** [comment]
Yes it was activated, after deactivating it everything is now fine, thanks.

But this raises some Questions:

Is this the default on Windows?
Should it really take that long?

**Comment by nirbheek** [comment]
> Is this the default on Windows?

It's not the default on Windows, no.

>  Should it really take that long?

Yes, if you enable that option and don't specify a Proxy-auto-config (PAC) URL, it searches a bunch of predefined URLs till it finds one.

libproxy should be caching that per-process, though, so only the *first* invocation of `px_proxy_factory_get_proxies()` should be slow. Subsequent ones should be fast. If that's not the case, that's the real bug here.

**Comment by lovetox** [comment]
> It's not the default on Windows, no.

OK thats good.

> libproxy should be caching that per-process, though, so only the _first_ invocation of `px_proxy_factory_get_proxies()` should be slow. Subsequent ones should be fast. If that's not the case, that's the real bug here.

I will test that and come back with the results. If its cached i think the 2 seconds on first call are alright.

**Comment by lovetox** [comment]
So i added a second call (see code below)

and Output is:

```
took 2.259612 seconds
second time took 2.252934 seconds
```


```
#include <glib.h>
#include <proxy.h>

int main()
{

  gint64 t1, t2, t3;

  pxProxyFactory *pf = px_proxy_factory_new();
  if (!pf) return 1;

  t1 = g_get_monotonic_time ();
  char **proxies = px_proxy_factory_get_proxies(pf, "https://www.google.com");
  t2 = g_get_monotonic_time ();

  proxies = px_proxy_factory_get_proxies(pf, "https://www.google.com");

  t3 = g_get_monotonic_time ();
  px_proxy_factory_free(pf);

  g_print ("took %f seconds\n", (gdouble) (t2 - t1) / 1000000);

  g_print ("second time took %f seconds", (gdouble) (t3 - t2) / 1000000);
  return 0;
}
```

**Comment by nirbheek** [comment]
> took 2.259612 seconds
> second time took 2.252934 seconds

Yeah looks like libproxy isn't caching which wapd addresses don't exist while searching for proxies. I don't have commit access to this project so I can't change the issue title, but I think this is enough investigation for now. Thanks!

**Comment by janbrummer** [comment]
Well i think the issue here is a different one. In case automatic proxy detection through wpad is enabled but there is no server providing the information it just fails. On the seconf try there is no existing data and it gives another try.
If there would be a wpad server the necessary values are filled and it skips the download part.

@DimStar77 Do you have an advice here? Just skip it after an error and wait for a network change does not sound right. As the server could have been offline for a short moment...
